Add goto issue id function

tyroneyeh opened the pull request:
for https://github.com/go-gitea/gitea/issues/4109#issuecomment-1527104992 

Supports format:
`#1234`
`Org/Repo#1234`

wxiaoguang commented:
I definitely like this idea more than adding the issue index into search engine (#19646)

Some more background: some people suggested to support the number without `#` prefix (the reasons is that many users do not have enough technical background).

And I guess we do not need to change the "search" behavior. So I think we can show a "goto" button dynamically like this, then all users could be satisfied.

```

[ input: #123 ]  [button:goto] [button:search] 

[ input: 1234 ]  [button:goto] [button:search] 

[ input:  Org/Repo#1234 ]  [button:goto] [button:search] 

[ input: other input ]  [button:search] 

```

tyroneyeh commented:
Added dynamic hash-tag button for go to function

![image](https://user-images.githubusercontent.com/1255041/235597090-76d20db4-2d86-4a8a-a18d-63f3d613339c.png)
![image](https://user-images.githubusercontent.com/1255041/235597106-fe67e845-9073-449f-8dcd-90b499e50b9b.png)

wxiaoguang commented:
Some quick reviews:

1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
2. According to https://docs.gitea.io/en-us/contributing/guidelines-frontend/
    * HTML ids and classes should use kebab-case
    * Show/hide: https://docs.gitea.io/en-us/contributing/guidelines-frontend/#showhide-elements
3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
4. The `aria-label=` is not accurate for the new button, it's not for "search", but for "goto"
5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)

silverwind commented:
Please use camelCase for variable names and don't be afraid to use longer names. I don't know what `pathslahnum` is supposed to mean.

silverwind commented on templates/repo/issue/search.tmpl:
```diff
@@ -8,7 +8,8 @@
 		<input type="hidden" name="poster" value="{{$.PosterID}}">
 		<input name="q" value="{{.Keyword}}" placeholder="{{.locale.Tr "explore.search"}}...">
-		<button class="ui small icon button" type="submit" aria-label="{{.locale.Tr "explore.search"}}">
+		<button id="hashtagbutton" class="ui small icon button" style="display:none" type="submit" aria-label="{{.locale.Tr "explore.search"}}">{{svg "octicon-hash"}}</button>
```
Use `gt-hidden` class to hide and then use `showElem`/`hideElemn`/`toggleElem` helper function from `utils/dom.js` to change visibility.

tyroneyeh commented:
> Some quick reviews:
> 
> 1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
Please help modify it, thanks

> 2. According to https://docs.gitea.io/en-us/contributing/guidelines-frontend/
>    
>    * HTML ids and classes should use kebab-case
>    * Show/hide: https://docs.gitea.io/en-us/contributing/guidelines-frontend/#showhide-elements
Modified

> 3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
Reuse submitted events

> 4. The `aria-label=` is not accurate for the new button, it's not for "search", but for "goto"
Modified

> 5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)
The goto button is hidden by default, and generally you don’t know that there is this function. It may be changed to the preset disabled, and it will be enabled after the input conforms to the format. It may be more intuitive.

wxiaoguang commented:
> > Some quick reviews:
> > 1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
> 
> Please help modify it, thanks
> 

Hmm ... I don't have writer permission at the moment (I quited manager team some time ago, just re-joint recently, so haven't got writer permission)

But `parseIssueHref` is covered by tests, the tests show how it works (if it helps)


> > 3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
> 
> Reuse submitted events
> 

I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?


> > 5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)
> 
> The goto button is hidden by default, and generally you don’t know that there is this function. It may be changed to the preset disabled, and it will be enabled after the input conforms to the format. It may be more intuitive.

"hidden by default" I think it is fine, adding a tooltip is good enough

wxiaoguang commented on templates/repo/issue/search.tmpl:
```diff
@@ -8,7 +8,8 @@
 		<input type="hidden" name="poster" value="{{$.PosterID}}">
 		<input name="q" value="{{.Keyword}}" placeholder="{{.locale.Tr "explore.search"}}...">
-		<button class="ui small icon button" type="submit" aria-label="{{.locale.Tr "explore.search"}}">
+		<button id="hashtag-button" class="ui small icon button gt-hidden" type="submit" aria-label="{{.locale.Tr "explore.goto"}}">{{svg "octicon-hash"}}</button>
```
Using `data-tooltip-content="{{.locale.Tr "explore.goto"}}"` is enough (no need to write aria-label)

wxiaoguang commented on web_src/js/features/repo-issue.js:
```diff
@@ -634,3 +634,23 @@ export function initRepoIssueBranchSelect() {
+    e.preventDefault();
+    window.location.href = !qval.includes('/') ? `${window.location.pathname}/${qval.replace('#', '')}` : `/${qval.replace('#', '/issues/')}`;
+  });
+  $('form.list-header-search input[name=q]').on('keyup', (e) => {
```
use `input` event , it works for mouse-pasting

tyroneyeh commented:
> I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?

I listen to the event processing sent by the form. If the form is not submit, there will be no action. If you enter #1234 in the input field and press enter, the submit event will also be trigger

wxiaoguang commented:
> > I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?
> 
> I listen to the event processing sent by the form. If the form is not submit, there will be no action. If you enter #1234 in the input field and press enter, the submit event will also be trigger

I see, you are right, I understand what the code does now. ps: all these `type=submit`  could be removed (but not a must), according to MDN: the buttons have `type=submit` by default 🤣  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

tyroneyeh commented:
> I see, you are right, I understand what the code does now. ps: all these `type=submit` could be removed (but not a must), according to MDN: the buttons have `type=submit` by default  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

ok, I removed it.

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -308,6 +308,7 @@ repos = Repositories
 users = Users
 organizations = Organizations
 search = Search
+goto = Goto
```
Would say make this `go_to = Go to`. "Goto" always makes me think of the same-named statement in some programming languages, so clear english seems better.

silverwind commented:
Counting slashes in the path seems to brittle. There is a rarely used method of installation where users install Gitea on a subdirectory, e.g. `https://example.com/gitea` and I assume this slash counting would break in such a scenario.

If possible re-use `parseIssueHref`  or add your own parser function in `utils.js`, with tests for it in `utils.test.js`.

tyroneyeh commented:
> Counting slashes in the path seems to brittle. There is a rarely used method of installation where users install Gitea on a subdirectory, e.g. `https://example.com/gitea` and I assume this slash counting would break in such a scenario.
> 
> If possible re-use `parseIssueHref` or add your own parser function in `utils.js`, with tests for it in `utils.test.js`.

I improved, please help me to check again
Thanks

lunny commented on web_src/js/features/repo-issue.js:
```diff
@@ -636,3 +637,29 @@ export function initRepoIssueBranchSelect() {
+    if (!$('#hashtag-button').length || aElm.id === 'search-button' || (aElm.name === 'q' && !qval.includes('#')) || (isGlobalIssuesArea && !qval.includes('/')) || !issueidre.test(qval)) return;
+    const {appUrl} = window.config;
+    const pathname = window.location.pathname;
+    const gotoUrl = qval.includes('/') ? `${appUrl}/${qval.replace('#', '/issues/')}` : `${appUrl}/${pathname}/${qval.replace('#', '')}`;
```
use `appSubUrl` directly.

lunny commented on web_src/js/features/repo-issue.js:
```diff
@@ -636,3 +637,29 @@ export function initRepoIssueBranchSelect() {
+    const qval = e.target.q.value;
+    const aElm = document.activeElement;
+    if (!$('#hashtag-button').length || aElm.id === 'search-button' || (aElm.name === 'q' && !qval.includes('#')) || (isGlobalIssuesArea && !qval.includes('/')) || !issueidre.test(qval)) return;
+    const {appSubUrl} = window.config;
```
I think it has been imported in line 9.

lunny commented on web_src/js/features/repo-issue.js:
```diff
@@ -636,3 +637,29 @@ export function initRepoIssueBranchSelect() {
+export function initRepoIssueGotoID() {
+  const issueidre = /^(?:\w+\/\w+#\d+|#\d+|\d+)$/;
+  const isGlobalIssuesArea = $('.repo.name.item').length > 0; // for global issues area or repository issues area
+  $('form.list-header-search').on('submit', (e) => {
```
class `list-header-search` has also been used on milestone search but only issues search are expected.

wxiaoguang reviewed (approved):
Some details do not look good to me (including: how the query text is parsed, when the button should be shown).

But I can't help to edit this PR. So, put my approval here.

After merging, I will propose new PRs for do more improvements.

lunny reviewed (approved):
The idea is good and the code could be improved by future PRs.

delvh commented:
I've just tested this functionality on try.gitea.io, and I have some reservations regarding it's current implementation:
1. It's not a good UI design to hide functionality like this completely. Typically, you should simply disable a button when it cannot be used (yet)
2. I am not a fan of detecting whether we are on the global issues page or not by checking if an element is present. This could have been a parameter, or some other form of detection that doesn't break as likely, i.e. using `window.config.pageData`.
3. The full `owner/repo#nr` syntax seems buggy: When I enter `delvh/test#3` on https://try.gitea.io/delvh/test/issues, for example, I get redirected to https://try.gitea.io/delvh/test/delvh/test/issues/3.
4. When I click on the `#` button and immediately return to the search page, the text will still be shown in the textarea **but** the button will only be shown when I manually change the text. We should also check if the initial text matches the pattern.

wxiaoguang commented:
> Some details do not look good to me (including: how the query text is parsed, when the button should be shown).
> But I can't help to edit this PR. So, put my approval here.
> After merging, I will propose new PRs for do more improvements.

Yup, it needs improvements

wxiaoguang commented:
->  Improve "goto issue by number" button #24577
